src/middleware.rs

```diff
@@ -58,11 +65,167 @@ use tracing::warn;
 ///     }
 ///
 ///     HttpResponse::Ok().body("Hello admin!")
-///   }
+/// }
 /// ```
 
 pub struct BiscuitMiddleware {
-    pub public_key: PublicKey,
+    public_key: Rc<dyn RootKeyProvider>,
+    error_handler: ErrorHandler,
+    token_extractor: TokenExtractor,
+}
+
+impl BiscuitMiddleware {
+    /// Instantiate a new middleware.
+    ///
+    /// [RootKeyProvider] is a [biscuit-auth](biscuit) trait used to delegate [PublicKey](biscuit::PublicKey) selection/rotation implementation.
+    ///
+    /// If either selection or rotation are not required, you can pass a [PublicKey](biscuit::PublicKey) (it already implements [RootKeyProvider] in [biscuit-auth](biscuit) lib).
+    ///
+    /// # Examples
+    /// Selection and rotation are not required
+    /// ```rust
+    /// use biscuit_actix_middleware::BiscuitMiddleware;
+    /// use biscuit_auth::KeyPair;
+    ///
+    /// BiscuitMiddleware::new(KeyPair::new().public());
+    /// ```
+    ///
+    /// Setup a simple KeyProvider that only accept token with a root_key_id corresponding to an index of an existing [PublicKey](biscuit_auth::PublicKey)
+    /// ```rust
+    /// use biscuit_auth::{error::Format, Biscuit, KeyPair, PublicKey, RootKeyProvider};
+    /// use biscuit_actix_middleware::BiscuitMiddleware;
+    ///
+    /// struct KeyProvider {
+    ///     roots: Vec<KeyPair>,
+    /// }
+    ///
+    /// impl KeyProvider {
+    ///     fn new() -> Self {
+    ///         let mut new = KeyProvider { roots: vec![] };
+    ///
+    ///         for _ in 1..11 {
+    ///             new.roots.push(KeyPair::new());
+    ///         }
+    ///
+    ///         new
+    ///     }
+    /// }
+    ///
+    /// impl RootKeyProvider for KeyProvider {
+    ///     fn choose(&self, key_id: Option<u32>) -> Result<PublicKey, Format> {
+    ///         Ok(self
+    ///             .roots
+    ///             .get(key_id.ok_or(Format::UnknownPublicKey)? as usize)
+    ///             .ok_or(Format::UnknownPublicKey)?
+    ///             .public())
+    ///     }
+    /// }
+    ///
+    /// let key_provider = KeyProvider::new();
+    ///
+    /// let mut builder = Biscuit::builder();
+    /// builder.set_root_key_id(2);
+    ///
+    /// let biscuit = builder.build(&key_provider.roots[2]).unwrap();
+    ///
+    /// assert!(Biscuit::from_base64(biscuit.to_base64().unwrap(), &key_provider).is_ok());
+    /// assert!(Biscuit::from_base64(biscuit.to_base64().unwrap(), &KeyPair::new().public()).is_err());
+    ///
+    /// let middleware = BiscuitMiddleware::new(key_provider);
+    /// ```
+    pub fn new<KP>(public_key: KP) -> BiscuitMiddleware
+    where
+        KP: RootKeyProvider + 'static,
+    {
+        BiscuitMiddleware {
+            public_key: Rc::new(public_key),
+            error_handler: |err: MiddlewareError, _: &ServiceRequest| err.error_response(),
+            token_extractor: |req: &ServiceRequest| {
+                Some(
```

Comment:
would it be possible to put this default behaviour in a named function? This would allow people to reuse it and maybe add a post-processing step if they want.

Reply:
Good idea! I guess you're talking about token_extractor one (as error_handle is pretty straightforward). About naming, what do you think about `token_extractor_default`?

Reply:
i think i have named it `default_token_extractor ` in haskell.

Reply:
Fix in 32ec78cbd2409dbe01a824097eb96e3c48884522